consul/autopilot.go

```diff
@@ -0,0 +1,298 @@
+type BasicAutopilot struct {
+	server *Server
+}
+
+// PromoteNonVoters promotes eligible non-voting servers to voters.
+func (b *BasicAutopilot) PromoteNonVoters(autopilotConf *structs.AutopilotConfig) error {
+	minRaftProtocol, err := ServerMinRaftProtocol(b.server.LANMembers())
+	if err != nil {
+		return fmt.Errorf("error getting server raft protocol versions: %s", err)
+	}
+
+	// If we don't meet the minimum version for non-voter features, bail early
+	if minRaftProtocol < 3 {
+		return nil
+	}
+
+	future := b.server.raft.GetConfiguration()
+	if err := future.Error(); err != nil {
+		return fmt.Errorf("failed to get raft configuration: %v", err)
+	}
+
+	var promotions []raft.Server
+	raftServers := future.Configuration().Servers
+	voterCount := 0
+	for _, server := range raftServers {
+		// If this server has been stable and passing for long enough, promote it to a voter
+		if server.Suffrage == raft.Nonvoter {
+			health := b.server.getServerHealth(string(server.Address))
+			if health != nil && health.Healthy && time.Now().Sub(health.StableSince) >= autopilotConf.ServerStabilizationTime {
+				promotions = append(promotions, server)
+			}
+		} else {
+			voterCount++
+		}
+	}
+
+	// Exit early if there's nothing to promote
+	if len(promotions) == 0 {
+		return nil
+	}
+
+	// If there's currently an even number of servers, we can promote the first server in the list
+	// to get to an odd-sized quorum
+	newServers := false
+	if voterCount%2 == 0 {
+		addFuture := b.server.raft.AddVoter(promotions[0].ID, promotions[0].Address, 0, 0)
+		if err := addFuture.Error(); err != nil {
+			return fmt.Errorf("failed to add raft peer: %v", err)
+		}
+		promotions = promotions[1:]
+		newServers = true
+	}
+
+	// Promote remaining servers in twos to maintain an odd quorum size
+	for i := 0; i < len(promotions)-1; i += 2 {
+		addFirst := b.server.raft.AddVoter(promotions[i].ID, promotions[i].Address, 0, 0)
+		if err := addFirst.Error(); err != nil {
+			return fmt.Errorf("failed to add raft peer: %v", err)
+		}
+		addSecond := b.server.raft.AddVoter(promotions[i+1].ID, promotions[i+1].Address, 0, 0)
+		if err := addSecond.Error(); err != nil {
+			return fmt.Errorf("failed to add raft peer: %v", err)
+		}
+		newServers = true
+	}
+
+	// If we added a new server, trigger a check to remove dead servers
+	if newServers {
+		select {
+		case b.server.autopilotRemoveDeadCh <- struct{}{}:
+		default:
+		}
+	}
+
+	return nil
+}
+
+// queryServerHealth fetches the raft stats for the given server and uses them
+// to update its ServerHealth
+func (s *Server) queryServerHealth(member serf.Member, server *agent.Server,
+	autopilotConf *structs.AutopilotConfig) (*structs.ServerHealth, error) {
+	stats, err := s.getServerStats(server)
+	if err != nil {
+		return nil, fmt.Errorf("error getting raft stats: %s", err)
+	}
+
+	health := &structs.ServerHealth{
+		ID:             server.ID,
+		Name:           server.Name,
+		SerfStatusRaw:  member.Status,
```

Comment:
The `Raw` status is a little weird. I'd probably carry just the Serf status and work on that internally, and send that out as a string to the outside world in the endpoint (we want to avoid a Serf dependency in the API client probably, too).